services/src/main/java/org/keycloak/userprofile/config/DeclarativeUserProfileModel.java

```diff
@@ -28,8 +28,8 @@
  */
 public class DeclarativeUserProfileModel extends ComponentModel {
 
-    public DeclarativeUserProfileModel() {
```

Comment:
How about just adding another constructor and delegate from the noargs ctor to the other one?

```java
...
public class DeclarativeUserProfileModel extends ComponentModel {

    public DeclarativeUserProfileModel() {
        this(DeclarativeUserProfileProvider.ID);
    }

    public DeclarativeUserProfileModel(String providerId) {
        setProviderId(providerId);
        setProviderType(UserProfileProvider.class.getName());
    }
}
```

Reply:
Thanks for the feedback, I really like the idea of the factory method, since it gives the possibility to use a complete different component model. I will enhance the pull request accordingly.

Regarding the default constructor I'm not yet 100% convinced.
Let me draft what the changes would mean for the developer of a custom user profile:

```java
public class DeclarativeUserProfileProvider extends ... {
  ...
  /* new factory method for creating compnent model
  *  uses default constructor -> model is hard coded in model class
  */
  protected ComponentModel createComponentModel(KeycloakSession session) {
     return new DeclarativeUserProfileModel();
 }
}


public class CustomUserProfileProvider extends DeclarativeUserProfileProvider {
  ...
  /*  has to be overriden in any sub class to ensure provider id is set correctly
  */
  protected ComponentModel createComponentModel(KeycloakSession session) {
     return new DeclarativeUserProfileModel(this.getId());
 }
}
```

When deriving from `DeclarativeUserProfileProvider` one would be enforced to override `createComponentModel()` in order to ensure that the provider id used in the component model is correct.

That wouldn't be required if the implementation in `DeclarativeUserProfileProvider`  would be like this:

```java
public class DeclarativeUserProfileProvider extends ... {
  ...
  /* new factory method for creating compnent model
  *  uses default constructor -> model is hard coded in model class
  */
  protected ComponentModel createComponentModel(KeycloakSession session) {
     return new DeclarativeUserProfileModel(this.getId());
 }
}

```
`getId()` needs to be overridden anyway since it is part of the `ProviderFactory` interface, so it would choose the correct id automatically.

It would still be possible to use a complete different implementation by overriding the factory method, but when nothing special is required it would just work as expected out-of-the-box.

What do you think about this @thomasdarimont ?

Reply:
@ioemat Your proposal looks better for me as it makes very clear the contract as well as a lot easier to extend the SPI.